plugins/amd-kria/fu-amd-kria-device.c

```diff
@@ -0,0 +1,252 @@
+{
+	FuAmdKriaDevice *self = FU_AMD_KRIA_DEVICE(device);
+	FuAmdKriaDevicePrivate *priv = GET_PRIVATE(self);
+	GUdevDevice *udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	const gchar *tmp;
+	g_auto(GStrv) of_name = NULL;
+
+	/* FuI2cDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_amd_kria_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/*
+	 * Fetch the OF_FULLNAME udev property and look for the I2C address in it
+	 * sample format: OF_FULLNAME=/axi/i2c@ff030000/eeprom@50
+	 */
+	tmp = g_udev_device_get_property(udev_device, "OF_FULLNAME");
+	if (tmp == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "no OF_FULLNAME property");
+		return FALSE;
+	}
+	of_name = fu_strsplit(tmp, strlen(tmp), "@", -1);
+	if (of_name == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "no '@' found in %s", tmp);
+		return FALSE;
+	}
+	tmp = of_name[g_strv_length(of_name) - 1];
+
+	if (g_strcmp0(priv->eeprom_address, tmp) != 0) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "invalid device");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_amd_kria_device_setup(FuDevice *device, GError **error)
+{
+	const gchar *devpath = fu_udev_device_get_sysfs_path(FU_UDEV_DEVICE(device));
+	FuAmdKriaDevice *self = FU_AMD_KRIA_DEVICE(device);
+	FuAmdKriaDevicePrivate *priv = GET_PRIVATE(self);
+	gsize bufsz = 0;
+	FuContext *ctx;
+	g_autofree gchar *buf = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *path = g_build_path("/", devpath, "eeprom", NULL);
+	g_autoptr(FuFirmware) firmware = NULL;
+	g_autoptr(GError) error_esp = NULL;
+	g_autoptr(GBytes) bytes = NULL;
+
+	if (!g_file_get_contents(path, &buf, &bufsz, error))
+		return FALSE;
+
+	/* parse the eeprom */
+	bytes = g_bytes_new(buf, bufsz);
+	firmware = fu_amd_kria_som_eeprom_new();
+	if (!fu_firmware_parse(firmware, bytes, FWUPD_INSTALL_FLAG_NONE, error))
+		return FALSE;
+
+	/* build instance IDs from EEPROM data */
+	fu_device_set_vendor(device, fu_amd_kria_som_get_manufacturer(FU_AMD_KRIA_SOM_EEPROM(firmware)));
+	vendor_id = g_strdup_printf("DMI:%s", fu_device_get_vendor(device));
+	fu_device_add_vendor_id(device, vendor_id);
+	fu_device_add_instance_str(device, "VENDOR", fu_device_get_vendor(device));
+	fu_device_add_instance_str(device, "PRODUCT", fu_amd_kria_som_get_product_name(FU_AMD_KRIA_SOM_EEPROM(firmware)));
+	fu_device_set_serial(device, fu_amd_kria_som_get_serial_number(FU_AMD_KRIA_SOM_EEPROM(firmware)));
+	if (!fu_device_build_instance_id(device, error, "KRIA", "VENDOR", NULL))
+		return FALSE;
+	if (!fu_device_build_instance_id(device, error, "KRIA", "VENDOR", "PRODUCT", NULL))
+		return FALSE;
+
+	/* setup the default ESP */
+	ctx = fu_device_get_context(device);
+	priv->esp = fu_context_get_default_esp(ctx, &error_esp);
+	if (priv->esp == NULL) {
```

Comment:
this feels like the wrong place to do this; maybe in ->constructed?

Reply:
You mean in the _plugin_ `constructed` callback?  I guess that means the device needs to pull data from the plugin then instead of the context.

Reply:
Also; I guess I'm not sure I see the benefit of ferrying the data around; it's a one device only plugin.

Reply:
Nah, in the GObject one, e.g. check out `plugins/vli/fu-vli-device.c`

Reply:
ahhh, I just hadn't noticed that callback existed :)

Fixed.